### PR TITLE
feat: add fuse shared expert configs for triton moe

### PR DESCRIPTION
**Pull request by froststeam**
## Summary
- Add Qwen3.5 MoE tuning entries with shared expert fusion enabled.
- Support auto dtype resolution from model config / quantization config.
- Add tuned MTT_S5000 configs for Qwen3.5 shared expert fusion cases.

## Test
```
root@worker35093:/data/torchada# python3 -m pytest tests/
=================================================================== test session starts ===================================================================
platform linux -- Python 3.10.12, pytest-9.0.3, pluggy-1.6.0
rootdir: /data/torchada
configfile: pyproject.toml
plugins: anyio-4.13.0
collected 374 items                                                                                                                                       

tests/test_cpp_extension.py ...............                                                                                                         [  4%]
tests/test_cuda_patching.py ..................................................................s.................................................... [ 35%]
...................................s.s.........................                                                                                     [ 52%]
tests/test_device_strings.py ................................                                                                                       [ 61%]
tests/test_extension_build.py ..sssssssss                                                                                                           [ 64%]
tests/test_graph_rotation.py .......                                                                                                                [ 66%]
tests/test_mappings.py ...................................................................................ssss                                      [ 89%]
tests/test_moe_config_dir.py .............                                                                                                          [ 92%]
tests/test_platform.py .........                                                                                                                    [ 95%]
tests/test_python_compat.py ........                                                                                                                [ 97%]
tests/test_stable_abi.py .                                                                                                                          [ 97%]
tests/test_stable_abi_mappings.py .......                                                                                                           [ 99%]
tests/test_triton_patching.py ..                                                                                                                    [100%]

==================================================================== warnings summary =====================================================================
tests/test_cpp_extension.py::TestCUDAExtension::test_create_extension_basic
tests/test_cpp_extension.py::TestCUDAExtension::test_create_extension_with_include_dirs
tests/test_cpp_extension.py::TestCUDAExtension::test_create_extension_with_extra_compile_args
  /usr/local/lib/python3.10/dist-packages/torch_musa/utils/musa_extension.py:803: UserWarning: TORCH_MUSA_ARCH_LIST is not set, all archs for visible cards are included for compilation. 
  If this is not desired, please set os.environ['TORCH_MUSA_ARCH_LIST'].
    warnings.warn(

tests/test_cuda_patching.py::TestLibraryImpl::test_library_impl_with_allow_override
  /usr/local/lib/python3.10/dist-packages/torch/library.py:364: UserWarning: Warning only once for all operators,  other operators may also be overridden.
    Overriding a previously registered kernel for the same operator and the same dispatch key
    operator: test_lib_725ebdab::test_op(Tensor x) -> Tensor
      registered at /data/torchada/tests/test_cuda_patching.py:1566
    dispatch key: CPU
    previous kernel: no debug info
         new kernel: registered at /data/torchada/tests/test_cuda_patching.py:1566 (Triggered internally at /home/pytorch/aten/src/ATen/core/dispatch/OperatorEntry.cpp:218.)
    self.m.impl(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================= 358 passed, 16 skipped, 4 warnings in 66.84s (0:01:06) ==================================================
```

### CODE REVIEW

**Comment by yeahdongcn** [comment]
@froststeam Check if the AI review comments need to be addressed or dropped.

File: src/torchada/triton/autotune/fused_moe/utils.py [hunk 1]
```diff
@@ -25,16 +32,112 @@ def calculate_shard_intermediate_size(
...
+def get_num_shared_experts(config, disable_shared_experts_fusion: bool) -> int:
+    if disable_shared_experts_fusion:
+        return 0
+    if getattr(config, "shared_expert_intermediate_size", None) is not None:
+        return 1
+    return getattr(config, "n_shared_experts", 0) or getattr(config, "num_shared_experts", 0) or 0
+
+
+def infer_quant_dtype_str(config) -> str:
+    quant_config = getattr(config, "quantization_config", None) or {}
```
[the remark this answers is not in the document]
Comment (froststeam) [reply src/torchada/triton/autotune/fused_moe/utils.py:53]: Fixed `dtype=auto` handling to require a real resolved dtype, and normalized benchmark dtype to `torch.dtype`.

File: src/torchada/triton/autotune/fused_moe/tune_moe.py [hunk 2]
```diff
@@ -426,6 +518,8 @@ def build_model_entries(args: argparse.Namespace) -> List[ModelEntry]:
             entry.hidden_size = params["hidden_size"]
             entry.shard_intermediate_size = params["shard_intermediate_size"]
             entry.topk = params["topk"]
+            entry.num_fused_shared_experts = params.get("num_fused_shared_experts", 0)
```
[the remark this answers is not in the document]
Comment (froststeam) [reply src/torchada/triton/autotune/fused_moe/tune_moe.py:550]: Fixed `get_num_shared_experts()` to prefer explicit shared expert count fields. `shared_expert_intermediate_size` is now only used as a Qwen-style single shared expert fallback, with a clarifying comment.

File: src/torchada/triton/runtime/fused_moe/router.py [hunk 3]
```diff
@@ -80,6 +80,36 @@ def scoring_func_impl(gating_output: torch.Tensor) -> torch.Tensor:
...
+    if routed_scaling_factor is not None:
+        shared_weights = topk_weights.sum(dim=-1, keepdim=True) / routed_scaling_factor
+    else:
+        shared_weights = topk_weights[:, :1]
+    shared_weights = shared_weights.expand(-1, num_fused_shared_experts)
+
+    return (
+        torch.cat([topk_weights, shared_weights], dim=-1),
+        torch.cat([topk_ids, shared_ids], dim=-1),
+    )
```
[the remark this answers is not in the document]
Comment (froststeam) [reply src/torchada/triton/runtime/fused_moe/router.py:112]: Fixed multi shared-expert weighting so total shared weight does not scale linearly with `num_fused_shared_experts`.

**Comment by froststeam** [comment]
> @froststeam Check if the AI review comments need to be addressed or dropped.

All review comments have been addressed.

**Review by yeahdongcn — APPROVED** [review]
